documentation/developing/modding_examples/blindfold_assist.mdx

```diff
@@ -0,0 +1,187 @@
+---
+sidebar_position: 0
+title: Blindfold Assist
+description: A deep dive into the code changes for the Blindfold Assist Mod
+---
+
+# Blindfold Assist Mod
+
+<PostAuthors authors={["dallmeyer"]} authorNames={["barg034"]}/>
+
+## Background
+
+Today I'm going to run through the process of building a mod to assist with blindfolded runs. I chose to kick things off with this mod because it only required a few simple changes, so I think reviewing the code should be easy enough. If you'd like to play this mod yourself, you can find it in the  [Unofficial Mod Launcher](https://opengoal-unofficial-mods.github.io/).
+
+![](./img/blindfold_assist/mod_image.png)
+
+If you're unfamiliar with [blindfolded video gaming](https://www.denofgeek.com/games/blindgolded-gaming-speedruns-history-explained/), one of the key steps is finding setups for progressing through levels that can be consistently repeated. In 3D games, you'll often see strategies like:
+- aligning the character against walls or other level geometry
+- being able to move in straight lines and/or consistent distances
+- manipulating the camera in consistent patterns
+
+When it comes to Jak, the game's responsive engine provides maybe too much flexibility in controlling Jak's movement and the camera 😅. This ends up hindering the consistency of blindfolded setups - so for this mod we'll actually be limiting some functionality! Here are the two changes I set out to tackle (thanks Borrdi for the suggestions):
+- limit analog stick input to 8 directions ↑ ↗ → ↘ ↓ ↙ ← ↖ (every 45 degrees)
+- add a button combo to (attempt to) center the camera behind Jak
+
+I'll go through things in more detail below, but you can [click here to see the full diff](https://github.com/OpenGOAL-Unofficial-Mods/blindfold-assist/commit/1cd217498b5efa501c4980241d253a6b09e8c330) of changes for this mod. It's only 69 lines across 3 files, and most of that is duplicated code for the left vs right analog sticks.
+
+<ReactPlayer controls url={require("./video/blindfold-assist.mp4").default} className="blog-video"/>
+
+## Limiting analog stick input to 8 directions
+
+### Where/how is controller input defined?
+
+On the OpenGOAL side, the controller input is represented by the [`cpad-info` type](https://github.com/open-goal/jak-project/blob/ee3bc182e13a440f90e20a6f7be7d3b294a7e56a/goal_src/jak1/engine/ps2/pad.gc#L49-L100). In particular, the below fields (some from the parent `hw-cpad` type) relate to the analog stick inputs:
+
+```lisp
+;; data that comes directly from hardware. it's 32 bytes + type tag (ignored in C kernel).
```

Comment:
you can actually use ```opengoal

It's not as good as the VSCode grammar because I basically had to re-write it and didn't spend as much time on it.  But it's still a little better than the lisp one (and maybe it'll get improved in the future)

There's also `opengoal_ir` and `opengoal_goos`